aiida/backends/tests/test_nodes.py

```diff
@@ -1702,100 +1621,385 @@ def test_deletion_non_existing_pk(self):
 
+        # By default deleting and input of the selection process will delete
+        # the workflow that involves the selection procedure but not any
+        # process that uses non-deleted inputs...
+        di1, di2, di3, di4, do1, pws, pcs, pwm = self._create_looped_graph()
+        uuids_check_existence = [n.uuid for n in (di1, di2, di3, do1, pcs)]
+        uuids_check_deleted = [n.uuid for n in (di4, pws, pwm)]
         with Capturing():
-            delete_nodes([wf.pk], verbosity=2, force=True, follow_calls=False, follow_returns=True)
+            delete_nodes([di4.pk], verbosity=2, force=True)
+        self._check_existence(uuids_check_existence, uuids_check_deleted)
 
+        # If the deleted input is the chosen one, then that's another story
+        di1, di2, di3, di4, do1, pws, pcs, pwm = self._create_looped_graph()
+        uuids_check_existence = [n.uuid for n in (di2, di3, di4)]
+        uuids_check_deleted = [n.uuid for n in (di1, do1, pws, pcs, pwm)]
+        with Capturing():
+            delete_nodes([di1.pk], verbosity=2, force=True)
         self._check_existence(uuids_check_existence, uuids_check_deleted)
+
+        # Deleting the workflow that chooses inputs should not touch the data
+        # provenance either.
+        di1, di2, di3, di4, do1, pws, pcs, pwm = self._create_looped_graph()
+        uuids_check_existence = [n.uuid for n in (di1, di2, di3, di4, do1, pcs)]
+        uuids_check_deleted = [n.uuid for n in (pws, pwm)]
+        with Capturing():
+            delete_nodes([pws.pk], verbosity=2, force=True)
         self._check_existence(uuids_check_existence, uuids_check_deleted)
 
-    def test_deletion_with_returns_n_loops(self):
-        """
-        Setting up a simple loop, to check that the following doesn't go bananas.
-        """
-        in1, in2, wf = [orm.Data().store(), orm.Data().store(), orm.WorkflowNode()]
-        wf.add_incoming(in1, link_type=LinkType.INPUT_WORK, link_label='link1')
-        wf.add_incoming(in2, link_type=LinkType.INPUT_WORK, link_label='link2')
-        wf.store()
-        in2.add_incoming(wf, link_type=LinkType.RETURN, link_label='link3')
+        # Again, deleting with call_calc/work_forward also deletes other parts of
+        # the graph.
+        di1, di2, di3, di4, do1, pws, pcs, pwm = self._create_looped_graph()
+        uuids_check_existence = [n.uuid for n in (di1, di2, di3, di4)]
+        uuids_check_deleted = [n.uuid for n in (do1, pws, pcs, pwm)]
+        with Capturing():
+            delete_nodes([pcs.pk], verbosity=2, force=True, call_calc_forward=True, call_work_forward=True)
+        self._check_existence(uuids_check_existence, uuids_check_deleted)
+
+    def _create_indep2w_graph(self):
+        """
+        Creates a simple graph with one workflow handling two independent workflows
+        (with one simple calculation each). It was designed and used mainly to point
+        out how the deletion behaviour works when setting call_calc_forward to true:
```

Comment:
```suggestion
        out how the deletion behaviour works when setting call_work_forward to true:
```
Also, add that if you set also `call_calc_forward` on top of this, actually also unrelated data will be deleted (DOB)